keps/sig-apps/20190318-PodDisruptionBudget-graduation-to-stable.md

```diff
@@ -0,0 +1,142 @@
+---
+title: graduate-pod-disruption-budget-to-stable
+authors:
+  - "@bsalamat"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "@liggitt"
+  - "@kow3ns"
+approvers:
+  - "@kow3ns"
+  - "@liggitt"
+editor: TBD
+creation-date: 2019-03-18
+last-updated: 2019-03-18
+status: implementable
+see-also:
+  - 
+replaces:
+superseded-by:
+---
+
+# Graduate PodDisruptionBudget to stable
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+[Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
+is a Kubernete API that limits the number of pods of a collection that are down simultaneously from voluntary disruptions.
+[Kubernetes eviction API](https://kubernetes.io/docs/tasks/administer-cluster/safely-drain-node/#the-eviction-api)
+takes PDB into account when terminating pods. If PDB is
+violated, the eviction API returns failure and does not delete the requested pod.
+The feature was introduced in Kubernetes 1.4 and promoted to beta in 1.5.
+It has been in beta for a long time. This document lays out the plan to promote
+it to stable.
+
+## Motivation
+
+PDB API has been stable and is an important feature that allows users to improve
+reliability of their critical workloads. This feature has been in beta for a
+long time. We need to promote to stable version given that we plan to support it
+long term.
+
+### Goals
+
+* Plan to promote PDB API to stable version.
+* Propose changes to the deployment controller to take PDB into account.
+* List action items to mitigate potential risks of supporting mutable PDB.
+
+### Non-Goals
+
+* Making changes to the API fields or their meanings.
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+#### Mutable PDBs
+
+A mutable PDB object allows its `MinAvailable` and `MaxUnavailable` fields to be
+modified by clients. Components that use PDB must watch such modifications and
+use the updated values when making decisions.
+
+This feature is implemented by [this PR](https://github.com/kubernetes/kubernetes/pull/69867).
+
+#### PDB in Deployment controller
+
+[Rolling update](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/rolling.go) of Deployment controller reads `MinAvailable` and `MaxUnavailable` from
+the rolling update spec and currently does not take PDB into account. We need to
+change the implementation and use
+`min(PDB.MinAvailable, RollingUpdate.MinAvailable)` and
+`max(PDB.MaxUnavailable, RollingUpdate.MaxUnavailable)` instead.
```

Comment:
Do we want deployments looking at PDB objects or using the evict subresource? Just looking at the PDB does not take into account PDB status for selectors that match across deployments, right?

Respecting PDB could easily lead to a situation in which a deployment could not make progress. Given that, is respecting PDB still desired? If so, describe how that state will be detected, communicated, and/or resolved (automatically or manually).

Reply:
What I had in mind was to only look at `MinAvailable` and `MaxUnavailable` values and treat those similar to `MinAvailable` and `MaxUnavailable` of a rolling update, without making further changes and without using "evict" subresource, but given your points and reading [some of the old comments](https://github.com/kubernetes/kubernetes/issues/12611#issuecomment-130506236), I think we should drop this requirement.